microsoft/PowerToys
# Add Delete functionality for Peek

## Summary of the Pull Request
This PR adds the ability for users to delete the file currently being previewed in Peek.


## PR Checklist

- [x] **Closes:** #22584
- [x] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: [#5168](https://github.com/MicrosoftDocs/windows-dev-docs/pull/5168)

## Detailed Description of the Pull Request / Additional comments
There are several new and changed elements in this PR:

- Navigation has been updated to track forward and backward movement through files. This is necessary so deleting a file always shows the next item in the natural sequence, i.e. the previous item if the user was moving backwards. This replicates Windows Photo and other media viewers.
- The indexes of deleted items are recorded rather than altering the `Items` collection (which could be very large). There are now `DisplayIndex` and `DisplayItemCount` properties, which reflect the calculated position and total. The old `CurrentIndex` still points to the actual index of the current item in the collection, but this is now internal only.
- The MainWindow ViewModel includes quite a bit of navigation-related code. We may wish to split this out into a separate model or provider in the future.
- When previewing a selection of items (where the number and total are shown in the title bar), these counts will always be shown. Previously, this info would not show if the count was 1. (This was likely because opening Peek while a single item is selected expands the Items collection to cover the entire folder and turns off the "(position/total)" display.)
- A text message has been added for when there are no more files remaining, i.e. the user has deleted the last one from their selection (or from the current folder if Peek wasn't opened via multi-selection). It may be worthwhile to add a simple icon by the side here - perhaps an image placeholder graphic? - but I've gone with the simple approach for now.
- Changes were made to the title bar code to ensure that the last filename and "open in application" details were not shown when there are no more files to preview. This was because there was an (entirely reasonable!) assumption in the code that there would always be one or more files being previewed.
- The delete operation sends the item to the Recycle Bin only. It was considered whether pressing `Shift + Delete` could do a permanent delete, and this is certainly possible in the future, but this release errs on the side of caution, especially as it's a new feature and involves deleting things.
- The Delete function itself relies upon a shell file operation, reflecting the fact that we're previewing shell items. This required adding some code to `NativeMethods.cs` for the API call, struct, flags, and error messages.
- Only *file* deletion is allowed. I considered allowing for empty folder deletes, too, but, again, this first release is deliberately conservative.
- Updated Peek documentation.

## Screenshots

### No More Files message
![Screenshot 2024-10-13 230651 (Small)](https://github.com/user-attachments/assets/3692460a-a209-4e57-9379-2d810d6e7d85)

### File Count For Single Remaining File
![image](https://github.com/user-attachments/assets/1e92e30f-bf45-45fd-8927-f119b1b29336)


## Validation Steps Performed

(Manual tests only.)

Tested:
- Single and multi-selection methods of opening Peek.
- Deleting from beginning, end and middle of a sequence (after opening Peek via multi-selection).
- Deleting until no files remained.
- Previewing a folder containing a mix of files and folders.
- Previewing a folder containing a mix of files which can be previewed and unsupported file types.
- Attempting to delete folders via Peek.
- Deleting a file in the selected list via Explorer before it could be previewed.
- Deleting a file from an external USB drive which doesn't support the Recycle Bin, to confirm that the system confirmation dialog is displayed which warns against permanent file deletion.
- Cancelling a permanent delete, confirming the file is re-added to the current list.
- Changing the "Don't show this warning again" setting in the delete confirmation dialog updates the settings.
- Updating the "Ask for confirmation before deleting files" setting in the Settings app is reflected in Peek immediately, and vice-versa. (The Settings app requires navigating to another page then returning to the Peek settings.)
